devices/uart-nrf9160/uart.c

```diff
@@ -0,0 +1,332 @@
+/*
+ * Phoenix-RTOS
+ *
+ * plo - operating system loader
+ *
+ * nRF9160 UART driver
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <lib/errno.h>
+#include <lib/lib.h>
+#include <devices/devs.h>
+
+#include <board_config.h>
+
+#define BUFFER_SIZE 0x20
+
+typedef struct {
+	volatile unsigned int *base;
+	unsigned int cnt;
+	unsigned int irq;
+
+	volatile char *tx_dma;
+	volatile char *rx_dma;
+
+	u8 dataRx[BUFFER_SIZE];
+	cbuffer_t cbuffRx;
+
+	u8 dataTx[BUFFER_SIZE];
+	cbuffer_t cbuffTx;
+} uart_t;
+
+
+struct {
+	uart_t uarts[UART_MAX_CNT];
+} uart_common;
+
+/* clang-format off */
+enum { cr1 = 0, cr2, cr3, brr, gtpr, rtor, rqr, isr, icr, rdr, tdr };
+/* clang-format on */
+
+static int uartLut[] = { UART0, UART1, UART2, UART3 };
+
+
+static const struct {
+	void *base;
+	unsigned int irq;
+	unsigned int baudrate;
+	volatile char *tx_dma;
+	volatile char *rx_dma;
+	unsigned char txpin;
+	unsigned char rxpin;
+	unsigned char rtspin;
+	unsigned char ctspin;
+} uartInfo[] = {
+	{ (void *)UART0_BASE, UART0_IRQ, UART_BAUDRATE, (void *)UART0_TX_DMA, (void *)UART0_RX_DMA,
+		UART0_TX, UART0_RX, UART0_RTS, UART0_CTS },
+	{ (void *)UART1_BASE, UART1_IRQ, UART_BAUDRATE, (void *)UART1_TX_DMA, (void *)UART1_RX_DMA,
+		UART1_TX, UART1_RX, UART1_RTS, UART1_CTS },
+	{ (void *)UART2_BASE, UART2_IRQ, UART_BAUDRATE, (void *)UART2_TX_DMA, (void *)UART2_RX_DMA,
+		UART2_TX, UART2_RX, UART2_RTS, UART2_CTS },
+	{ (void *)UART3_BASE, UART3_IRQ, UART_BAUDRATE, (void *)UART3_TX_DMA, (void *)UART3_RX_DMA,
+		UART3_TX, UART3_RX, UART3_RTS, UART3_CTS }
+};
+
+
+static uart_t *uart_getInstance(unsigned int minor)
+{
+	if (minor >= UART_MAX_CNT) {
+		return NULL;
+	}
+
+	if (uartLut[minor] == 0) {
+		return NULL;
+	}
+
+	return &uart_common.uarts[minor];
+}
+
+
+static int uart_handleIntr(unsigned int irq, void *buff)
+{
+	uart_t *uart = (uart_t *)buff;
+
+	if (uart == NULL) {
+		return -EINVAL;
+	}
+
+	if (uart->base + uarte_events_rxdrdy) {
+		/* clear rxdrdy event flag */
+		*(uart->base + uarte_events_rxdrdy) = 0u;
+		/* we discard volatile type intentionally */
+		lib_cbufWrite(&uart->cbuffRx, &((char *)uart->rx_dma)[uart->cnt], 1);
```

Comment:
I would do it like that to avoid `discarding volatile` warning: `lib_cbufWrite(&uart->cbuffRx, (char *)uart->rx_dma + uart->cnt, 1);`

Reply:
just adding my 0.03 USD, casting away `const` and `volatile` directly breaks MISRA required rule 11.8 (_A cast shall not remove any const or volatile qualification from the type pointed to by a pointer_).

Also, using `memcpy` on `volatile` memory is not safe. Not sure if `hal_memcpy` is volatile-friendly TBH.

Reply:
@nalajcie Strictly speaking it's not - it does not respect access size, i.e. it can access data by both words or bytes. This is in theory not acceptable for volatiles.

Reply:
So using `char temp;` like that would be ok here?

```C
		if (uart->cnt < UART_RX_DMA_SIZE) {
			temp = *(uart->rx_dma + uart->cnt);
			lib_cbufWrite(&uart->cbuffRx, &temp, 1);
			uart->cnt++;
		}
```

Reply:
@damianloew yes

Reply:
ok, already implemented.